python/lsst/resources/http.py

```diff
@@ -313,13 +313,35 @@ def _as_local(self) -> Tuple[str, bool]:
         temporary : `bool`
             Always returns `True`. This is always a temporary file.
         """
-        log.debug("Downloading remote resource as local file: %s", self.geturl())
+        log.debug(f"Downloading {self.geturl()} as local file")
         r = self.session.get(self.geturl(), stream=True, timeout=TIMEOUT)
         if r.status_code != 200:
             raise FileNotFoundError(f"Unable to download resource {self}; status code: {r.status_code}")
-        with tempfile.NamedTemporaryFile(suffix=self.getExtension(), delete=False) as tmpFile:
+
+        # Compute the block size to the file system where temporary files are
+        # to be written. Use the locations pointed to by the environment
+        # variables used by tempfile.mkstemp() to determine the paths of the
+        # temporary directories.
+        buffering = -1
+        chunk_size = None
+        for path in (os.getenv(var) for var in ('TMPDIR', 'TEMP', 'TMP')):
```

Comment:
Why are we guessing where the file is going to be written? We know where it is going to be written.

Reply:
> Why are we guessing where the file is going to be written? We know where it is going to be written.

These are the locations where `tempfile.mkstemp()` creates the temporary file, if any of them are defined. At this moment in the execution, it is not known which one will be selected, so we have to inspect them in order.

Reply:
This is an excerpt of the documentation of [tempfile.mkstemp](https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp)

> _If dir is not None, the file will be created in that directory; otherwise, a default directory is used. The default directory is chosen from a platform-dependent list, but the user of the application can control the directory location by setting the TMPDIR, TEMP or TMP environment variables._ 

Should we then enforce, at the application level, that the directory for temporary file creation must be pointed to `TMPDIR` (which is a POSIX-defined variable) the dfault value being `/tmp` ? I find that a cleaner solution than to create every temporary file twice just to retrieve the block size and could even retrieve that block size once and save it.

Reply:
If we believe that the block size will be fixed for the entire process (modulo shenanigans of tests redefining TMPDIR) then can't we add a class method for doing this where we will create a throwaway temp file once and cache the result?

I suppose another approach is to use the `dir` parameter and define exactly where we want the file to end up -- we can use `tmpdir = os.environ.get("TMPDIR", ResourcePath(".", forceAbsolute=True, forceDirectory=True))` (or do the loop you do now with `cwd` at the end)  and then examine `tmpdir` directly. Still might be worth caching this rather than doing the fsstat call every time.

```python
_TMPDIR = None
def _get_temp_dir():
    if _TMPDIR is None:
        for tmpdir in (...):
            ...
        _TMPDIR = (tmpdir, buffering)
    return _TMPDIR

def _as_local(self):
    tmpdir, buffering = _get_temp_dir()
    ...
```
or something?